From ec6d002e0f8b09f4580a97bfa1706baec52ec85a Mon Sep 17 00:00:00 2001
From: Christopher Degawa <ccom@randomderp.com>
Date: Fri, 11 Jun 2021 15:18:51 -0500
Subject: [PATCH 2/6] FindTIFF: Use pkg-config for finding dependencies

libtiff if built statically often relies on additional libraries such
as libjpeg, libdeflate, and zstd

Signed-off-by: Christopher Degawa <ccom@randomderp.com>
---
 Modules/FindTIFF.cmake | 36 ++++++++++++++++++++++++++++++++----
 1 file changed, 32 insertions(+), 4 deletions(-)

diff --git a/Modules/FindTIFF.cmake b/Modules/FindTIFF.cmake
index ed2657c4ab..030d78b413 100644
--- a/Modules/FindTIFF.cmake
+++ b/Modules/FindTIFF.cmake
@@ -68,9 +68,29 @@ The following cache variables may also be set:
 cmake_policy(PUSH)
 cmake_policy(SET CMP0057 NEW)
 
-find_path(TIFF_INCLUDE_DIR tiff.h)
+function(_FIND_TIFF_PKG_CONFIG)
+  find_package(PkgConfig)
+  if(NOT PKG_CONFIG_FOUND)
+      return()
+  endif()
+  pkg_check_modules(TIFF QUIET libtiff-4)
+  if(NOT TIFF_FOUND)
+      return()
+  endif()
+  set(TIFF_FOUND ${TIFF_FOUND} PARENT_SCOPE)
+  set(TIFF_INCLUDE_DIR ${TIFF_INCLUDEDIR} PARENT_SCOPE)
+  set(TIFF_INCLUDE_DIRS ${TIFF_INCLUDE_DIRS} PARENT_SCOPE)
+  set(TIFF_LIBRARIES ${TIFF_LIBRARIES} PARENT_SCOPE)
+  list(GET TIFF_LINK_LIBRARIES 0 TIFF_LIBRARY)
+  set(TIFF_LIBRARY ${TIFF_LIBRARY} PARENT_SCOPE)
+endfunction()
 
 set(TIFF_NAMES ${TIFF_NAMES} tiff libtiff tiff3 libtiff3)
+
+_FIND_TIFF_PKG_CONFIG()
+
+find_path(TIFF_INCLUDE_DIR tiff.h)
+
 foreach(name ${TIFF_NAMES})
   list(APPEND TIFF_NAMES_DEBUG "${name}d")
 endforeach()
@@ -127,13 +147,14 @@ FIND_PACKAGE_HANDLE_STANDARD_ARGS(TIFF
                                   VERSION_VAR TIFF_VERSION_STRING)
 
 if(TIFF_FOUND)
-  set(TIFF_LIBRARIES ${TIFF_LIBRARY})
+  list(APPEND TIFF_INCLUDE_DIRS ${TIFF_INCLUDE_DIR})
+  list(REMOVE_DUPLICATES TIFF_INCLUDE_DIRS)
+  list(APPEND TIFF_LIBRARIES ${TIFF_LIBRARY})
+  list(REMOVE_DUPLICATES TIFF_LIBRARIES)
   if("CXX" IN_LIST TIFF_FIND_COMPONENTS AND NOT MSVC)
     list(APPEND TIFF_LIBRARIES ${TIFFXX_LIBRARY})
   endif()
 
-  set(TIFF_INCLUDE_DIRS "${TIFF_INCLUDE_DIR}")
-
   if(NOT TARGET TIFF::TIFF)
     add_library(TIFF::TIFF UNKNOWN IMPORTED)
     if(TIFF_INCLUDE_DIRS)
@@ -159,6 +180,13 @@ if(TIFF_FOUND)
         IMPORTED_LINK_INTERFACE_LANGUAGES_DEBUG "C"
         IMPORTED_LOCATION_DEBUG "${TIFF_LIBRARY_DEBUG}")
     endif()
+    if(TIFF_LDFLAGS)
+      target_link_options(TIFF::TIFF INTERFACE ${TIFF_LDFLAGS})
+      target_link_libraries(TIFF::TIFF INTERFACE ${TIFF_LDFLAGS})
+    endif()
+    if(TIFF_CFLAGS)
+      target_compile_options(TIFF::TIFF INTERFACE ${TIFF_CFLAGS})
+    endif()
   endif()
 
   if(NOT TARGET TIFF::CXX)
-- 
2.31.1

